torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2431,54 +2431,90 @@ Tensor linalg_qr_backward(const std::vector<torch::autograd::Variable> &grads, c
   }
 }
 
-// Invertible case is derived from Jacobi's formula, and also can be found at:
-// http://eprints.maths.ox.ac.uk/1079/1/NA-08-01.pdf
 Tensor linalg_det_backward(const Tensor & grad, const Tensor& self, const Tensor& det) {
-  auto singular_case_backward = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
-    Tensor u, sigma, v;
-    std::tie(u, sigma, v) = self.svd();
-    auto gsigma = prod_backward(grad.unsqueeze(-1), sigma, det.unsqueeze(-1));
-    return svd_backward({{}, gsigma, {}}, self, true, true, u, sigma, v);
+  if (self.numel() == 0) {
+    return at::empty_like(self);
+  }
+  auto det_backward_nonsingular = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
+    // Derived from Jacobi's formula for partial derivative, which can be found
+    // at https://en.wikipedia.org/wiki/Jacobi%27s_formula
+
+    // Expand each scalar determinant into a diagonal matrix whose size matches
+    // the matrices in `self`. This is required for at::solve
+    auto det_sizes = det.sizes().vec();
+    auto det_strides = det.strides().vec();
+    det_sizes.push_back(self.size(-1));
+    det_strides.push_back(0);
+    auto det_vecs = det.as_strided(det_sizes, det_strides);
+    auto det_diag = det_vecs.diag_embed();
+    auto det_grad = std::get<0>(at::solve(det_diag, self.transpose(-1, -2))).conj();
+
+    return unsqueeze_multiple(grad, {-1, -2}, self.dim()) * det_grad;
   };
 
-  auto nonsingular_case_backward = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
-    return unsqueeze_multiple(grad * det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1);
+  auto det_backward_singular = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
+    // Derived from the gradient formula that would be used if `self`'s
+    // determinant is calculated using SVD, like so:
+    //    u, s, v = svd(self)
+    //    det(self) = det(u) * prod(s) * det(v).conj()
+    //
+    // This formula should be correct even if `self` is nonsingular.
+    Tensor u, s, v;
+    std::tie(u, s, v) = self.svd();
+    auto u_det = u.det();
+    auto s_prod = s.prod(-1);
+    auto v_det = v.det();
+
+    auto v_det_grad = grad.conj() * (u_det * s_prod);
+    auto v_grad = det_backward_nonsingular(v_det_grad, v, v_det);
+
+    auto u_det_times_s_prod_grad = grad * v_det;
+
+    auto u_det_grad = u_det_times_s_prod_grad * s_prod.conj();
+    auto u_grad = det_backward_nonsingular(u_det_grad, u, u_det);
+
+    auto s_prod_grad = handle_r_to_c(s_prod.scalar_type(), u_det_times_s_prod_grad * u_det.conj());
+    auto s_grad = prod_backward(s_prod_grad, s, s_prod, -1, false);
+
+    return svd_backward({u_grad, s_grad, v_grad}, self, true, true, u, s, v);
   };
 
+  double singular_det_cutoff = 1e-8;
+
```

Comment:
Since we have an access to the full SVD, we could set the cutoff based on the largest singular value. You could have a look at how `matrix_rank` method does that.

Reply:
We could, but isn't SVD pretty slow? The nonsingular case might slow down significantly

Reply:
You are right, I forgot that we do not compute the SVD for non-singular cases. 
But we can probably upper bound the largest singular value, which, I believe, could be better than just having a hardcoded value.

For example, the paper https://core.ac.uk/download/pdf/82616334.pdf states that for a squared matrix `X` of size `n x n` we have:
```
||X||_2 <= 2 / |det(X)| (||X||_F / sqrt(n + 1))^{n + 1}
```
Since we have already `det(X)` computed, we can use this formula as a surrogate for the largest singular value. The rest could be done similar to what `matrix_rank` is doing.

What are your thoughts on this, @mruberry , @Lezcano , @IvanYashchuk ?

Reply:
That bound seems way too large to be practical IMO. For what is worth, the (operator) norm 2 of a matrix is not expensive to compute via the power method (I assume that's what `torch.linalg.norm` does, I do not know).

That being said, I think that a better approach here would be to use that the derivatives of the determinant are continuous. The singular matrices form a set of measure zero (in particular it is closed), so any continuous formula to approximate the derivative of the gradient for non-singular matrices should to the correct value in the boundary (i.e. the singular matrices). Have you tried if the formula for the non-singular case alone passes the `gradcheck` tests?

On the point above, note that you are solving the system: AX = det(A) I. This `X` is literally the (transpose of the) adjugate, so I believe that this formula should work just fine for both cases (again, modulo numerical errors / stability).

Reply:
@Lezcano, unfortunately `solve` doesn't work for when the input is singular:

```python
>>> a = torch.tensor([[ 20., -40., -35., -40.],
...         [ -4.,   8.,   7.,   8.],
...         [-28.,  56.,  49.,  56.],
...         [-36.,  72.,  63.,  72.]])
>>> a_det = a.det()
>>> a_det
tensor(0.)
>>> a_det_diag = a_det * torch.eye(4)
>>> torch.linalg.solve(a.t(), a_det_diag)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: linalg_solve: U(2,2) is zero, singular U.
```

Reply:
It should (hopefully) work if and only if `torch.inverse` (or `A.inv()`) works.

In any case, I would not worry too much, as @nikitaved and I have been discussing the implementation of a more efficient backwards formula. Hopefully we should be able to ship it for 1.9.

Reply:
Alright, so is it ok if we leave it as it is in this PR?

Reply:
Looks like there's a lot of background on this issue, maybe we should take it offline for a linalg sync discussion where we can all get together on VC?

We have one later this week

Reply:
@kurtamohler  I think we can leave it as it is in this PR.

@mruberry and yes, we can discuss this one in the linalg sync this week.

Reply:
Once `lu_unpack` is merged we can quickly mend `det_backward`.

Reply:
If we need a VC, I'm happy to join in. It seems like we're clear on the path forward though

Reply:
I invited you to the sing tomorrow morning, @kurtamohler, if you'd like to join, but if you, @nikitaved and @Lezcano are comfortable with a plan then no worries

Reply:
I am happy with the plan as it is. We were testing @nikitaved and I some implementations and they pass `gradcheck`, so I think we should be good.